src/Microsoft.NetCore.Analyzers/Core/Tasks/DoNotCreateTaskCompletionSourceWithWrongArguments.cs

```diff
@@ -0,0 +1,89 @@
+                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                // Ensure necessary types are available
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTaskCompletionSource, out var tcsType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemObject, out var objectType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTaskContinuationOptions, out var taskContinutationOptionsType))
+                {
+                    return;
+                }
+
+                compilationContext.RegisterOperationAction(operationContext =>
+                {
+                    // Ensure that you're using the TaskCompletionSource constructor with the object state property else exit
+                    var invocation = (IObjectCreationOperation)operationContext.Operation;
+                    if (!invocation.Type.OriginalDefinition.Equals(tcsType) ||
+                        invocation.Arguments.Length == 0 ||
+                        !invocation.Constructor.Parameters[0].Type.Equals(objectType))
+                    {
+                        return;
+                    }
+
+                    // Check to see if argument is a conversion operator
+                    if (!(invocation.Arguments[0] is IArgumentOperation arg) ||
+                        arg.Value.Kind != OperationKind.Conversion ||
+                        !(arg.Value is IConversionOperation conversionOperation))
+                    {
+                        return;
+                    }
+
+                    // Check to see if its an enum field or local
+                    if (conversionOperation.Operand.Kind != OperationKind.FieldReference &&
+                        conversionOperation.Operand.Kind != OperationKind.LocalReference)
+                    {
+                        return;
+                    }
+
+                    // If the operand is TaskContinuationOptions then report diagnostic
+                    if (!conversionOperation.Operand.Type.Equals(taskContinutationOptionsType))
+                    {
+                        return;
+                    }
+
+                    operationContext.ReportDiagnostic(Diagnostic.Create(Rule, conversionOperation.Syntax.GetLocation()));
```

Comment:
It seems that your chosen diagnostic message needs an argument for the first parameter name of the invoked constructor

```suggestion
                    operationContext.ReportDiagnostic(conversionOperation.CreateDiagnostic(Rule, invocation.Arguments[0].Parameter.Name));
```